src/concurrency/transaction_context.cpp

```diff
@@ -80,102 +78,59 @@ void TransactionContext::Init(const size_t thread_id,
+  rw_set_[location] = RWType::READ;
 }
 
 void TransactionContext::RecordReadOwn(const ItemPointer &location) {
-  auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    RWType rw_type = rw_set_it->second;
-    PELOTON_ASSERT(rw_type != RWType::DELETE && rw_type != RWType::INS_DEL);
-    if (rw_type == RWType::READ) {
-      rw_set_it->second = RWType::READ_OWN;
-    }
-  } else {
-    rw_set_.insert(rw_set_it, std::make_pair(location, RWType::READ_OWN));
-  }
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
+  rw_set_[location] = RWType::READ_OWN;
 }
 
 void TransactionContext::RecordUpdate(const ItemPointer &location) {
-  auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    RWType rw_type = rw_set_it->second;
-    if (rw_type == RWType::READ || rw_type == RWType::READ_OWN) {
-      is_written_ = true;
-      rw_set_it->second = RWType::UPDATE;
-    } else if (rw_type == RWType::UPDATE || rw_type == RWType::INSERT) {
-      return;
-    } else {
-      // DELETE or INS_DELETE
-      PELOTON_ASSERT(false);
-    }
-  } else {
-    rw_set_.insert(rw_set_it, std::make_pair(location, RWType::UPDATE));
-  }
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
+  rw_set_[location] = RWType::UPDATE;
 }
 
 void TransactionContext::RecordInsert(const ItemPointer &location) {
-  auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    PELOTON_ASSERT(false);
-    return;
-  }
-  rw_set_.insert(rw_set_it, std::make_pair(location, RWType::INSERT));
-  ++insert_count_;
+  PELOTON_ASSERT(rw_set_.count(location) == 0);
+  rw_set_[location] = RWType::INSERT;
 }
 
-bool TransactionContext::RecordDelete(const ItemPointer &location) {
+void TransactionContext::RecordDelete(const ItemPointer &location) {
+  PELOTON_ASSERT(rw_set_.count(location) == 0 ||
+                 (rw_set_[location] != RWType::DELETE &&
+                  rw_set_[location] != RWType::INS_DEL));
   auto rw_set_it = rw_set_.find(location);
-  if (rw_set_it != rw_set_.end()) {
-    RWType rw_type = rw_set_it->second;
-
-    if (rw_type == RWType::READ || rw_type == RWType::READ_OWN) {
-      rw_set_it->second = RWType::DELETE;
-      is_written_ = true;
-      return false;
-    } else if (rw_type == RWType::UPDATE) {
-      rw_set_it->second = RWType::DELETE;
-      return false;
-    } else if (rw_type == RWType::INSERT) {
-      rw_set_it->second = RWType::INS_DEL;
-      --insert_count_;
-      return true;
-    } else {
-      // DELETE and INS_DEL
-      PELOTON_ASSERT(false);
-      return false;
-    }
+  if (rw_set_it != rw_set_.end() && rw_set_it->second == RWType::INSERT) {
+    rw_set_it->second = RWType::INS_DEL;
```

Comment:
I am just wondering - what does happen, if the value of `rw_set_it->second` changes between line 130 and line 131/133. Do we still do the right thing?

Reply:
I'll have to test. I suspect this entire logic may be susceptible to races and we might need to go back to the drawing board to think about thread-safe RWSet semantics.